File: 1106 Parsing A Boolean Expression/1106.cpp
class Solution {
public:
    void reduce(stack<char>& stk){
        bool a = true;
        bool b = false;

        while (stk.top() != '('){
            bool cur = stk.top() == 't' ? true : false;
            stk.pop();
            a &= cur;
            b |= cur;
        }
        stk.pop();

        char op = stk.top();
        stk.pop();

        if (op == '|'){
            a = b;
        } else if (op == '!') {
            a = !a;
        }

        stk.push(a ? 't' : 'f');
    }

    bool parseBoolExpr(string expression) {
        stack<char> stk;

        for (char c: expression){
            if (c == ','){
                continue;
            } else if (c != ')'){
                stk.push(c);
            } else {
                reduce(stk);
            }
        }

        return stk.top() == 't';
    }
};

File: 1106 Parsing A Boolean Expression/1106.py
class Solution:
    def parseBoolExpr(self, expression: str) -> bool:
        stk = []

        def reduce():
            values = []

            while stk[-1] != '(':
                values.append(stk.pop())
            stk.pop() # remove '('
            
            op = stk.pop()

            if op == '&':
                stk.append(all(values))
            elif op == '|':
                stk.append(any(values))
            else:
                stk.append(not values[0])

        for c in expression:
            if c == ',':
                continue
            elif c != ')':
                if c == 't':
                    stk.append(True)
                elif c == 'f':
                    stk.append(False)
                else:
                    stk.append(c)
            else:
                reduce()
        
        return stk.pop()
